In [1]:
!pip install geocoder
!pip install folium

     |████████████████████████████████| 102kB 7.4MB/s ta 0:00:011
     |████████████████████████████████| 102kB 5.5MB/s ta 0:00:011     |█████████████████▌              | 51kB 3.8MB/s eta 0:00:01


### Importing libraries

In [2]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print("All Required Libraries Imported!")

All Required Libraries Imported!


### 2. Data Extraction and Cleaning

In [4]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
extracting_data = requests.get(url).text
wiki_data = BeautifulSoup(extracting_data, 'lxml')

Converting content of PostalCode HTML table as dataframe

In [5]:
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

content = wiki_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,'Borough': borough,'Neighborhood': neighborhood},ignore_index=True)

In [6]:
toronto = toronto[toronto.Borough!='Not assigned']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1

In [7]:
df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head()

Postalcode         Borough                            Neighborhood
0      M1A\n  Not assigned\n                          Not assigned\n
1      M1B\n   Scarborough\n                          Malvern, Rouge
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill
4      M1G\n   Scarborough\n                                  Woburn

In [8]:
df.describe()

Postalcode         Borough    Neighborhood
count         180             180             180
unique        180              11             100
top         M6K\n  Not assigned\n  Not assigned\n
freq            1              77              77

In [9]:
df = df.dropna()
empty = 'Not assigned'
df = df[(df.Postalcode != empty ) & (df.Borough != empty) & (df.Neighborhood != empty)]

In [10]:
df.head()

Postalcode         Borough                            Neighborhood
0      M1A\n  Not assigned\n                          Not assigned\n
1      M1B\n   Scarborough\n                          Malvern, Rouge
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill
4      M1G\n   Scarborough\n                                  Woburn

In [11]:
def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df_2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')

In [12]:

df_2.describe()

Postalcode         Borough    Neighborhood
count         180             180             180
unique        180              11             100
top         M6K\n  Not assigned\n  Not assigned\n
freq            1              77              77

In [13]:
print(df_2.shape)
df_2.head()

(180, 3)


Postalcode         Borough                            Neighborhood
0      M1A\n  Not assigned\n                          Not assigned\n
1      M1B\n   Scarborough\n                          Malvern, Rouge
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill
4      M1G\n   Scarborough\n                                  Woburn

In [14]:
def get_latilong(postal_code):
    lati_long_coords = None
    while(lati_long_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(postal_code))
        lati_long_coords = g.latlng
    return lati_long_coords
    
get_latilong('M4G')

[43.70909000000006, -79.36409999999995]

In [15]:
postal_codes = df_2['Postalcode']    
coords = [ get_latilong(postal_code) for postal_code in postal_codes.tolist() ]

In [16]:
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])
df_2['Latitude'] = df_coords['Latitude']
df_2['Longitude'] = df_coords['Longitude']

In [17]:
df_2[df_2.Postalcode == 'M5G']

Empty DataFrame
Columns: [Postalcode, Borough, Neighborhood, Latitude, Longitude]
Index: []

In [18]:
df_2.head(10)

Postalcode         Borough                                     Neighborhood  \
0      M1A\n  Not assigned\n                                   Not assigned\n   
1      M1B\n   Scarborough\n                                   Malvern, Rouge   
2      M1C\n   Scarborough\n           Rouge Hill, Port Union, Highland Creek   
3      M1E\n   Scarborough\n                Guildwood, Morningside, West Hill   
4      M1G\n   Scarborough\n                                           Woburn   
5      M1H\n   Scarborough\n                                        Cedarbrae   
6      M1J\n   Scarborough\n                              Scarborough Village   
7      M1K\n   Scarborough\n      Kennedy Park, Ionview, East Birchmount Park   
8      M1L\n   Scarborough\n                  Golden Mile, Clairlea, Oakridge   
9      M1M\n   Scarborough\n  Cliffside, Cliffcrest, Scarborough Village West   

   Latitude  Longitude  
0  43.64869  -79.38544  
1  43.81153  -79.19552  
2  43.78564  -79.15871  
3  43.76575  -79.17520  
4  43.76820  -79.21761  
5  43.76969  -79.23944  
6  43.74309  -79.23526  
7  43.72861  -79.26367  
8  43.71406  -79.28412  
9  43.72360  -79.23496

In [19]:
address = 'Scarborough,Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude_x = location.latitude
longitude_y = location.longitude
print('The Geograpical Co-ordinate of Seattle,Washington are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Seattle,Washington are 43.773077, -79.257774.


### 3. Map of Scarborough

In [20]:
map_Scarborough = folium.Map(location=[latitude_x, longitude_y], zoom_start=10)

for lat, lng, nei in zip(df_2['Latitude'], df_2['Longitude'], df_2['Neighborhood']):
    
    label = '{}'.format(nei)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Scarborough)  
    
map_Scarborough

In [21]:
address = 'Scarborough,Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude_n1 = location.latitude
longitude_n1 = location.longitude
print('The Geograpical Co-ordinate of Neighborhood_1 are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Neighborhood_1 are 43.773077, -79.257774.


In [22]:
CLIENT_ID = 'RWBH522MUIUB4TXVCQL0PFZCZ325Q2Q4WUZ5WUSBZBUBB2Y5' 
CLIENT_SECRET = 'PNJDBMRDII5ZSHTHYIDT1X2NK5U02FCULLGU0PU0AC3QF55E'
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: '+CLIENT_ID)
print('CLIENT_SECRET: '+CLIENT_SECRET)

Your credentails:
CLIENT_ID: RWBH522MUIUB4TXVCQL0PFZCZ325Q2Q4WUZ5WUSBZBUBB2Y5
CLIENT_SECRET: PNJDBMRDII5ZSHTHYIDT1X2NK5U02FCULLGU0PU0AC3QF55E


In [23]:
radius = 700 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude_n1, 
   longitude_n1, 
    radius, 
   LIMIT)
results = requests.get(url).json()


In [24]:
venues=results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)
nearby_venues.columns

Index(['reasons.count', 'reasons.items', 'referralId', 'venue.categories',
       'venue.id', 'venue.location.address', 'venue.location.cc',
       'venue.location.city', 'venue.location.country',
       'venue.location.crossStreet', 'venue.location.distance',
       'venue.location.formattedAddress', 'venue.location.labeledLatLngs',
       'venue.location.lat', 'venue.location.lng',
       'venue.location.neighborhood', 'venue.location.postalCode',
       'venue.location.state', 'venue.name', 'venue.photos.count',
       'venue.photos.groups', 'venue.venuePage.id'],
      dtype='object')

In [25]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### 4. Nearby Venues/Locations

In [26]:
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()

venue.name  \
0                   SEPHORA   
1              Disney Store   
2      American Eagle Store   
3                 DAVIDsTEA   
4  St. Andrews Fish & Chips   

                                    venue.categories  venue.location.lat  \
0  [{'id': '4bf58dd8d48988d10c951735', 'name': 'C...           43.775017   
1  [{'id': '4bf58dd8d48988d1f3941735', 'name': 'T...           43.775537   
2  [{'id': '4bf58dd8d48988d103951735', 'name': 'C...           43.776012   
3  [{'id': '4bf58dd8d48988d1dc931735', 'name': 'T...           43.776320   
4  [{'id': '4edd64a0c7ddd24ca188df1a', 'name': 'F...           43.771865   

   venue.location.lng  
0          -79.258109  
1          -79.256833  
2          -79.258334  
3          -79.258688  
4          -79.252645

### 5. Categories of Nearby Venues/Locations

In [27]:
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(5)

name         categories        lat        lng
0                   SEPHORA     Cosmetics Shop  43.775017 -79.258109
1              Disney Store   Toy / Game Store  43.775537 -79.256833
2      American Eagle Store     Clothing Store  43.776012 -79.258334
3                 DAVIDsTEA           Tea Room  43.776320 -79.258688
4  St. Andrews Fish & Chips  Fish & Chips Shop  43.771865 -79.252645

In [28]:
a=pd.Series(nearby_venues.categories)
a.value_counts()[:10]

Clothing Store        8
Coffee Shop           5
Restaurant            4
Tea Room              2
Pharmacy              2
Sandwich Place        2
Gas Station           2
Deli / Bodega         1
Shopping Mall         1
Italian Restaurant    1
Name: categories, dtype: int64

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=700):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # making GET request
        venue_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
Scarborough_venues = getNearbyVenues(names=df_2['Neighborhood'],
                                   latitudes=df_2['Latitude'],
                                   longitudes=df_2['Longitude']
                                  )

Not assigned

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Parkwoods
Don Mills
Don Mills
Not assigned

Not assigned

Bathurst Manor, Wilson Heights, Downsview North
No

In [31]:
print('There are {} Uniques Categories.'.format(len(Scarborough_venues['Venue Category'].unique())))
Scarborough_venues.groupby('Neighborhood').count().head()

There are 302 Uniques Categories.


Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                           30   
Alderwood, Long Branch                                               9   
Bathurst Manor, Wilson Heights, Downsview North                      4   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   27   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                            30   
Alderwood, Long Branch                                                9   
Bathurst Manor, Wilson Heights, Downsview North                       4   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    27   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                           30              30   
Alderwood, Long Branch                               9               9   
Bathurst Manor, Wilson Heights, Downsview North      4               4   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   27              27   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                     30   
Alderwood, Long Branch                                         9   
Bathurst Manor, Wilson Heights, Downsview North                4   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             27   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                    30  
Alderwood, Long Branch                                        9  
Bathurst Manor, Wilson Heights, Downsview North               4  
Bayview Village                                               4  
Bedford Park, Lawrence Manor East                            27

### One Hot Encoding of Features

In [32]:
Scarborough_onehot = pd.get_dummies(Scarborough_venues[['Venue Category']], prefix="", prefix_sep="")


Scarborough_onehot['Neighborhood'] = Scarborough_venues['Neighborhood'] 


fixed_columns = [Scarborough_onehot.columns[-1]] + list(Scarborough_onehot.columns[:-1])
Scarborough_onehot = Scarborough_onehot[fixed_columns]
Scarborough_grouped = Scarborough_onehot.groupby('Neighborhood').mean().reset_index()
Scarborough_onehot.head(5)

Zoo Exhibit  ATM  Accessories Store  Afghan Restaurant  African Restaurant  \
0            0    0                  0                  0                   0   
1            0    0                  0                  0                   0   
2            0    0                  0                  0                   0   
3            0    0                  0                  0                   0   
4            0    0                  0                  0                   0   

   Airport  American Restaurant  Animal Shelter  Antique Shop  Aquarium  \
0        0                    0               0             0         0   
1        0                    0               0             0         0   
2        0                    0               0             0         0   
3        0                    0               0             0         0   
4        0                    0               0             0         0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 1   

   Athletics & Sports  Auto Dealership  Auto Garage  Auto Workshop  \
0                   0                0            0              0   
1                   0                0            0              0   
2                   0                0            0              0   
3                   0                0            0              0   
4                   0                0            0              0   

   Automotive Shop  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  \
0                0          0           0                0           0   
1                0          0           0                0           0   
2                0          0           0                0           0   
3                0          0           0                0           0   
4                0          0           0                0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  \
0       0     0    0               0                 0                   0   
1       0     0    0               0                 0                   0   
2       0     0    0               0                 0                   0   
3       0     0    0               0                 0                   0   
4       0     0    0               0                 0                   0   

   Beach  Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0      0          0         0           0                   0          0   
1      0          0         0           0                   0          0   
2      0          0         0           0                   0          0   
3      0          0         0           0                   0          0   
4      0          0         0           0                   0          0   

   Bistro  Board Shop  Bookstore  Botanical Garden  Boutique  \
0       0           0          0                 0         0   
1       0           0          0                 0         0   
2       0           0          0                 0         0   
3       0           0          0                 0         0   
4       0           0          0                 0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridge  Bubble Tea Shop  \
0                     0               0        0       0                0   
1                     0               0        0       0                0   
2                     0               0        0       0                0   
3                     0               0        0       0                0   
4                     0               0        0       0                0   

   Buildin

In [33]:
num_top_venues = 5
for hood in Scarborough_grouped['Neighborhood']:
    print("---- "+hood+" ----")
    temp =Scarborough_grouped[Scarborough_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Agincourt ----
                venue  freq
0       Shopping Mall  0.10
1  Chinese Restaurant  0.10
2         Pizza Place  0.07
3     Badminton Court  0.03
4              Bakery  0.03


---- Alderwood, Long Branch ----
               venue  freq
0  Convenience Store  0.11
1        Pizza Place  0.11
2        Gas Station  0.11
3                Gym  0.11
4     Sandwich Place  0.11


---- Bathurst Manor, Wilson Heights, Downsview North ----
                  venue  freq
0                  Park  0.50
1  Other Great Outdoors  0.25
2     Convenience Store  0.25
3           Music Store  0.00
4          Optical Shop  0.00


---- Bayview Village ----
              venue  freq
0           Dog Run  0.25
1              Park  0.25
2             Trail  0.25
3  Asian Restaurant  0.25
4       Music Venue  0.00


---- Bedford Park, Lawrence Manor East ----
                venue  freq
0          Restaurant  0.07
1      Sandwich Place  0.07
2     Thai Restaurant  0.07
3         Pizza Place  0.07
4  It

In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Most Common venues near neighborhood

In [35]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Scarborough_grouped['Neighborhood']

for ind in np.arange(Scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt         Shopping Mall   
1                           Alderwood, Long Branch                  Pool   
2  Bathurst Manor, Wilson Heights, Downsview North                  Park   
3                                  Bayview Village               Dog Run   
4                Bedford Park, Lawrence Manor East           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0    Chinese Restaurant           Pizza Place  Latin American Restaurant   
1              Pharmacy        Sandwich Place          Convenience Store   
2  Other Great Outdoors     Convenience Store                Yoga Studio   
3      Asian Restaurant                  Park                      Trail   
4        Sandwich Place           Coffee Shop                 Restaurant   

  5th Most Common Venue 6th Most Common Venue    7th Most Common Venue  \
0       Badminton Court                  Bank           Discount Store   
1           Coffee Shop                   Gym              Gas Station   
2     Electronics Store               Dog Run         Doner Restaurant   
3           Yoga Studio     Electronics Store         Doner Restaurant   
4       Thai Restaurant    Italian Restaurant  Comfort Food Restaurant   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0      Sushi Restaurant           Supermarket                       Bakery  
1           Pizza Place                   Pub                   Donut Shop  
2            Donut Shop   Dumpling Restaurant  Eastern European Restaurant  
3            Donut Shop   Dumpling Restaurant  Eastern European Restaurant  
4          Intersection  Fast Food Restaurant                     Pharmacy

### K-Means Clustering Approach

In [36]:
Scarborough_grouped_clustering = Scarborough_grouped.drop('Neighborhood', 1)
kmeans = KMeans(n_clusters=3, random_state=0).fit(Scarborough_grouped_clustering)
kmeans.labels_

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 2], dtype=int32)

In [37]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Scarborough_merged =df_2.iloc[:16,:]


Scarborough_merged = Scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Scarborough_merged.head()

Postalcode         Borough                            Neighborhood  \
0      M1A\n  Not assigned\n                          Not assigned\n   
1      M1B\n   Scarborough\n                          Malvern, Rouge   
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek   
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill   
4      M1G\n   Scarborough\n                                  Woburn   

   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.64869  -79.38544               1           Coffee Shop   
1  43.81153  -79.19552               1           Zoo Exhibit   
2  43.78564  -79.15871               1         Moving Target   
3  43.76575  -79.17520               0                  Park   
4  43.76820  -79.21761               0  Fast Food Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Hotel                  Café   Japanese Restaurant   
1       Paintball Field  Fast Food Restaurant        Farmers Market   
2     Fish & Chips Shop                   Bar           Event Space   
3    Athletics & Sports  Gym / Fitness Center           Yoga Studio   
4                  Park           Coffee Shop    Chinese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar           Pizza Place    Italian Restaurant   
1     Electronics Store               Dog Run      Doner Restaurant   
2      Doner Restaurant            Donut Shop   Dumpling Restaurant   
3     Electronics Store               Dog Run      Doner Restaurant   
4           Yoga Studio  Ethiopian Restaurant      Doner Restaurant   

         8th Most Common Venue 9th Most Common Venue  \
0                    Bookstore            Steakhouse   
1                   Donut Shop   Dumpling Restaurant   
2  Eastern European Restaurant     Electronics Store   
3                   Donut Shop   Dumpling Restaurant   
4                   Donut Shop   Dumpling Restaurant   

        10th Most Common Venue  
0                          Bar  
1  Eastern European Restaurant  
2         Ethiopian Restaurant  
3  Eastern European Restaurant  
4  Eastern European Restaurant

### Map of Clusters

In [39]:
kclusters = 10

In [40]:
map_clusters = folium.Map(location=[latitude_x, longitude_y], zoom_start=11)


x = np.arange(kclusters)
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]
print(rainbow)


markers_colors = []
for lat, lon, nei , cluster in zip(Scarborough_merged['Latitude'], 
                                   Scarborough_merged['Longitude'], 
                                   Scarborough_merged['Neighborhood'], 
                                   Scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(nei) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

['#8000ff', '#4856fb', '#10a2f0', '#2adddd', '#62fbc4', '#9cfba4', '#d4dd80', '#ffa256', '#ff562c', '#ff0000']


In [41]:
df1=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 0,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df2=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 1,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df3=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 2,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]

In [44]:
Scarborough_Avg_HousingPrice = pd.DataFrame({"Neighborhood":df_2["Neighborhood"],
                                      "Average_Housing_Price":[335000.0,286600.0,175000.0,225900.0,219400.0,
                                                               573900.0,225000.0,370500.0,370500.0,433500.0,279200.0,
                                                               279200.0,225000.0,370500.,255400.0,433500.0,433500.0,
                                                               435000.0,289500.0,265000.0,285900.0,239400.0,
                                                               589900.0,295000.0,380500.0,378500.0,438500.0,229200.0,
                                                               229200.0,365000.0,388500.,285400.0,493500.0,477500.0,378000.0,316600.0,195000.0,225900.0,219400.0,
                                                               573900.0,367000.0,370500.0,370500.0,363500.0,279200.0,
                                                               279200.0,271000.0,370500.,255400.0,383500.0,433500.0,335000.0,286600.0,185000.0,225900.0,219400.0,
                                                               573900.0,329000.0,370500.0,370500.0,533500.0,279200.0,
                                                               279200.0,375000.0,370500.,255400.0,493500.0,433500.0,335000.0,286600.0,165000.0,225900.0,219400.0,
                                                               573900.0,425000.0,370500.0,370500.0,433500.0,279200.0,
                                                               279200.0,195000.0,370500.,255400.0,403500.0,433500.0,335000.0,286600.0,187000.0,225900.0,219400.0,
                                                               573900.0,325000.0,370500.0,370500.0,333500.0,279200.0,
                                                               279200.0,289000.0,370500.,255400.0,413500.0,433500.0,254800.0
                                                              
                                                              
                                                              ]})

ValueError: array length 103 does not match index length 180

In [45]:
Scarborough_Avg_HousingPrice.set_index('Neighborhood',inplace=True,drop=True)

NameError: name 'Scarborough_Avg_HousingPrice' is not defined

In [46]:
Scarborough_Avg_HousingPrice.plot(kind='bar',figsize=(24,18),alpha=0.75)

NameError: name 'Scarborough_Avg_HousingPrice' is not defined

In [47]:
clusters=pd.DataFrame({"Cluster1":df1["Neighborhood"],
                      "Cluster2":df2["Neighborhood"],
                      "Cluster4":df3["Neighborhood"]})
clusters = clusters.replace(np.nan, '', regex=True)

In [48]:
new_Scarborough=Scarborough_merged.set_index("Neighborhood",drop=True)
#Source:https://www.greatschools.org
Scarborough_school_ratings=pd.DataFrame({"Neighborhood":df["Neighborhood"],
                                      "Top School Rating":[7,9,5,8,10,10,7,10,1,2,1,2,7,2,3,2,6,
                                                           5,4,8,9,9,6,6,4,5,4,6,8,10,8,9,6,2,
                                                           10,2,5,8,9,6,6,10,8,9,1,2,3,4,5,6,9,
                                                           8,5,9,6,9,6,4,8,10,2,5,6,3,9,8,7,
                                                           7,8,5,8,9,1,5,4,7,2,3,6,6,9,4,8,7,
                                                           4,8,9,2,6,4,7,5,10,4,6,8,9,7,5,6,5,8,7
                                                           ]})

ValueError: array length 103 does not match index length 180

Conclusion: In this project, using k-means cluster algorithm I separated the neighborhood into 10(Ten) different clusters and for 103 different lattitude and logitude from dataset, which have very-similar neighborhoods around them. Using the charts above results presented to a particular neighborhood based on average house prices and school rating have been made.